nuhal_tiva/src/pin_tiva.c

```diff
@@ -200,7 +244,76 @@ bool pin_read(uint32_t pin)
     const uint32_t base = pin_base(pin);
     const uint32_t mask = pin_mask(pin);
     return HWREG(base + (GPIO_O_DATA + (mask << 2)));
+}
+
+void analog_pin_read(uint32_t* pins, size_t num_pins, bool use_adc_1, uint16_t* out)
+{
+    if (num_pins > 8)
+    {
+        error(FILE_LINE, "Too many pins to analog read from");
+    }
+
+    uint32_t bases[8] = {0};
+    uint32_t masks[8] = {0};
+    uint8_t ains[8] = {0};
+    uint32_t adc_raw[8] = {0};
+
+     // options for analog pin:
+    // - option 1: Using ADC0 or ADC1
+    // - option 2: Using SS0 (sample up to 8 times sequentially), SS1 (up to 4), SS2 (up to 4), OR SS3 (only 1)
+    // - option 3: How many pins are we reading from sequentially and from which ain channel?
+    
+    // Option 1: ADC0 or ADC1
+    uint32_t adcsysctl = use_adc_1 ? SYSCTL_PERIPH_ADC1 : SYSCTL_PERIPH_ADC0;
+    uint32_t adcbase   = use_adc_1 ? ADC1_BASE : ADC0_BASE;
 
+    // Option 2: find appropriate sample sequencer for number of pins that will be read from
+    uint8_t ss = 0;
+    if (num_pins > 4)
+    {
+        ss = 0; // SS0 supports up to 8 steps
+    }
+    else if (num_pins > 1)
+    {
+        ss = 1; // SS1 supports up to 4 steps
+    }
+    else
+    {
+        ss = 3; // SS3 supports 1 step
+    }
+
+    tiva_peripheral_enable(adcsysctl);
+
+    //Configure Sample Sequencer to use given AIN channel
+    ADCSequenceDisable(adcbase, ss);  // Disable SS for config
+    ADCSequenceConfigure(adcbase, ss, ADC_TRIGGER_PROCESSOR, 0); // Set software trigger and priority
+    for (size_t step = 0; step < num_pins; step++) // for option 3: configure to read from selected pins
+    {
+        bases[step] = pin_base(pins[step]);
+        masks[step] = pin_mask(pins[step]);
+        ains[step] = get_ain_channel(bases[step], masks[step]); // map pins to corresponding AIN channels
+        
+        uint32_t config = ADC_CTL_CH0 + (ains[step] & 0xF);
+        if (step == num_pins - 1)
+        {
+            config |= ADC_CTL_IE | ADC_CTL_END;
+        }
+
+        ADCSequenceStepConfigure(adcbase, ss, step, config);
+    }
+    ADCSequenceEnable(adcbase, ss); // Enable SS back on
+    ADCIntClear(adcbase, ss); // Clear any interrupt flags
+
+    ADCProcessorTrigger(adcbase, ss);                // Trigger SS conversion
+    while(!ADCIntStatus(adcbase, ss, false));        // Wait for conversion completion
```

Comment:
Never run an infinite loop without the ability to timeout. In this case the timeout can be set based on how long the read is expected to take, + a little margin of error.